File: 07-security/06-ssrf/nodejs/ssrf.js
// Layer 7 · Topic 6 — SSRF: validate the connection, not the string (Node.js).
//
// One command, no deps, no network: `node ssrf.js`.
// The README's Node path: undici's Agent accepts a custom `connect` option --
// the correct place to pin, because you get the hostname and return a socket,
// so validation and connection happen in the same step with no window between
// them. `fetch` follows redirects by DEFAULT (redirect: 'follow'); 'manual' is
// the right setting for a fetcher. Node has no built-in private-range check, so
// this file implements one -- which is exactly the code a real `connect` hook
// needs.
//
// The finding: string_blocklist ALLOWs every internal target via an encoding
// it did not enumerate; resolve_and_pin BLOCKs them by checking the resolved IP.

const FAKE_DNS = {
  'internal-admin': '10.7.0.10',
  metadata: '10.7.0.169',
  'allowed.test': '93.184.216.34',
  'a.rebind.lab.test': '10.7.0.10',
  localhost: '127.0.0.1',
};
const STRING_DENY = ['localhost', '127.0.0.1', '169.254.169.254'];
const PAYLOADS = [
  ['http://internal-admin:8000/secrets', 'plain internal reach'],
  ['http://10.7.0.169/latest/meta-data/iam/...', 'cloud metadata / credential theft'],
  ['http://0/secrets', '0 == 0.0.0.0 (this host)'],
  ['http://2130706433/', 'decimal form of 127.0.0.1'],
  ['http://[::1]:8000/', 'IPv6 loopback'],
  ['http://ok.test@10.7.0.10/secrets', 'userinfo confusion (real host after @)'],
  ['http://a.rebind.lab.test/secrets', 'DNS rebinding (TOCTOU)'],
];

function hostOf(raw) {
  try {
    const h = new URL(raw).hostname;      // drops userinfo
    return h.replace(/^\[|\]$/g, '');      // strip IPv6 brackets
  } catch { return ''; }
}

function canonicalIP(host) {
  if (FAKE_DNS[host]) host = FAKE_DNS[host];
  if (/^\d+$/.test(host)) {                // "0", "2130706433"
    const n = Number(host) >>> 0;
    return `${(n >>> 24) & 255}.${(n >>> 16) & 255}.${(n >>> 8) & 255}.${n & 255}`;
  }
  if (/^\d+\.\d+\.\d+\.\d+$/.test(host) || host.includes(':')) return host;
  return null; // unknown name
}

function isDenied(ip) {
  if (ip.includes(':')) { // IPv6
    const l = ip.toLowerCase();
    return l === '::1' || l === '::' || l.startsWith('fe80') || l.startsWith('fc') || l.startsWith('fd');
  }
  const [a, b] = ip.split('.').map(Number);
  return a === 127 || a === 10 || a === 0 ||
    (a === 172 && b >= 16 && b <= 31) ||
    (a === 192 && b === 168) ||
    (a === 169 && b === 254);
}

function verdictBlocklist(raw) {
  // A real naive blocklist runs over the URL TEXT as sent. (Note: new URL()
  // would canonicalize the decimal IP to 127.0.0.1 and accidentally catch it
  // -- an interesting WHATWG-parser nuance, but not what a string filter does.)
  const low = raw.toLowerCase();
  return STRING_DENY.some((d) => low.includes(d)) ? 'BLOCK' : 'ALLOW';
}
function verdictResolvePin(raw) {
  const ip = canonicalIP(hostOf(raw));
  if (ip === null) return ['BLOCK', 'unresolvable'];
  return [isDenied(ip) ? 'BLOCK' : 'ALLOW', ip];
}

console.log('Layer 7 · Topic 6 — SSRF: string blocklist vs resolve-and-pin\n');
console.log(`   ${'payload'.padEnd(44)}${'blocklist'.padEnd(11)}${'resolve+pin'.padEnd(13)}resolved`);
let rb = 0, rp = 0;
for (const [url] of PAYLOADS) {
  const v1 = verdictBlocklist(url);
  const [v2, ip] = verdictResolvePin(url);
  if (v1 === 'ALLOW') rb++;
  if (v2 === 'ALLOW') rp++;
  console.log(`   ${url.padEnd(44)}${v1.padEnd(11)}${v2.padEnd(13)}${ip}`);
}
console.log(`\n   internal targets reached -- string_blocklist: ${rb}/${PAYLOADS.length}   resolve_and_pin: ${rp}/${PAYLOADS.length}`);
console.log('\nIMDS v1 vs v2: v1 returns credentials to a plain GET; v2 refuses a GET');
console.log('without a PUT-obtained token -> 0 bytes. v2 raises the bar, not the fix.');
console.log('\nRead: the STRING is not the ADDRESS. Pin in undici\'s connect hook so');
console.log('validation and the socket use the same resolved IP -- no rebinding window.');

File: 07-security/06-ssrf/java/Ssrf.java
// Layer 7 · Topic 6 — SSRF: validate the connection, not the string (Java).
//
// Runs on the JDK alone: `javac Ssrf.java && java Ssrf`.
// The README's Java notes: java.net.http.HttpClient defaults to Redirect.NEVER
// (the SAFE default, the opposite of Python's requests), and the historical
// hazard is that the legacy java.net.URL does DNS resolution inside equals()
// and hashCode() -- so a validation cache keyed on URL can be poisoned by DNS.
// Use java.net.URI for anything not being fetched right now. This program uses
// URI to extract the host and InetAddress to classify the resolved literal.
//
// The finding: string_blocklist ALLOWs every internal target via an encoding
// it did not enumerate; resolve_and_pin BLOCKs them by classifying the IP.
import java.net.InetAddress;
import java.net.URI;
import java.util.LinkedHashMap;
import java.util.Map;

public class Ssrf {
    static final Map<String, String> FAKE_DNS = Map.of(
        "internal-admin", "10.7.0.10",
        "metadata", "10.7.0.169",
        "allowed.test", "93.184.216.34",
        "a.rebind.lab.test", "10.7.0.10",
        "localhost", "127.0.0.1");
    static final String[] STRING_DENY = {"localhost", "127.0.0.1", "169.254.169.254"};

    static String hostOf(String raw) {
        try {
            String h = URI.create(raw).getHost();
            if (h == null) return "";
            return h.replaceAll("^\\[|\\]$", ""); // strip IPv6 brackets
        } catch (Exception e) { return ""; }
    }

    static String canonicalIP(String host) {
        String h = FAKE_DNS.getOrDefault(host, host);
        if (h.matches("\\d+")) {                     // "0", "2130706433"
            long n = Long.parseLong(h) & 0xFFFFFFFFL;
            return String.format("%d.%d.%d.%d", (n >> 24) & 255, (n >> 16) & 255, (n >> 8) & 255, n & 255);
        }
        if (h.matches("\\d+\\.\\d+\\.\\d+\\.\\d+") || h.contains(":")) return h;
        return null; // unknown name
    }

    static boolean isDenied(String ip) {
        try {
            InetAddress a = InetAddress.getByName(ip); // literal IP -> no DNS
            return a.isLoopbackAddress() || a.isSiteLocalAddress()
                || a.isLinkLocalAddress() || a.isAnyLocalAddress();
        } catch (Exception e) { return true; } // fail closed
    }

    static String verdictBlocklist(String raw) {
        String low = raw.toLowerCase();
        for (String d : STRING_DENY) if (low.contains(d)) return "BLOCK";
        return "ALLOW";
    }

    static String[] verdictResolvePin(String raw) {
        String ip = canonicalIP(hostOf(raw));
        if (ip == null) return new String[]{"BLOCK", "unresolvable"};
        return new String[]{isDenied(ip) ? "BLOCK" : "ALLOW", ip};
    }

    public static void main(String[] args) {
        Map<String, String> payloads = new LinkedHashMap<>();
        payloads.put("http://internal-admin:8000/secrets", "plain internal reach");
        payloads.put("http://10.7.0.169/latest/meta-data/iam/...", "credential theft");
        payloads.put("http://0/secrets", "0 == 0.0.0.0");
        payloads.put("http://2130706433/", "decimal 127.0.0.1");
        payloads.put("http://[::1]:8000/", "IPv6 loopback");
        payloads.put("http://ok.test@10.7.0.10/secrets", "userinfo confusion");
        payloads.put("http://a.rebind.lab.test/secrets", "DNS rebinding");

        System.out.println("Layer 7 · Topic 6 — SSRF: string blocklist vs resolve-and-pin\n");
        System.out.printf("   %-44s%-11s%-13s%s%n", "payload", "blocklist", "resolve+pin", "resolved");
        int rb = 0, rp = 0;
        for (String url : payloads.keySet()) {
            String v1 = verdictBlocklist(url);
            String[] v2 = verdictResolvePin(url);
            if (v1.equals("ALLOW")) rb++;
            if (v2[0].equals("ALLOW")) rp++;
            System.out.printf("   %-44s%-11s%-13s%s%n", url, v1, v2[0], v2[1]);
        }
        System.out.printf("%n   internal targets reached -- string_blocklist: %d/%d   resolve_and_pin: %d/%d%n",
            rb, payloads.size(), rp, payloads.size());
        System.out.println("\nIMDS v1 vs v2: v1 returns credentials to a plain GET; v2 refuses without");
        System.out.println("a PUT-obtained token -> 0 bytes. v2 raises the bar, it is not the fix.");
        System.out.println("\nRead: the STRING is not the ADDRESS. Classify the RESOLVED IP; keep URI");
        System.out.println("out of hash-based caches, because java.net.URL resolves DNS in equals().");
    }
}
